src/framework/constraints.ml

```diff
@@ -243,9 +243,37 @@ struct
     if M.tracing then M.traceu "combine" "combined: %a" S.D.pretty r;
     r
 
-  let tf_special_call man lv f args = S.special man lv f args
 
-  let tf_proc var edge prev_node lv e args getl sidel demandl getg sideg d =
+  let rec tf_proc var edge prev_node lv e args getl sidel demandl getg sideg d =
+    let tf_special_call man f =
+      let once once_control init_routine =
+        (* Executes leave event for new local state d if it is not bottom *)
+        let leave_once d =
+          if not (S.D.is_bot d) then
+            let rec man' =
+              { man with
+                ask = (fun (type a) (q: a Queries.t) -> S.query man' q);
+                local = d;
+              }
+            in
+            S.event man' (Events.LeaveOnce { once_control }) man'
+          else
+            S.D.bot ()
+        in
+        let first_call =
+          let d' = S.event man (Events.EnterOnce { once_control;  ran = false }) man in
+          tf_proc var edge prev_node None init_routine [] getl sidel demandl getg sideg d'
+        in
+        let later_call = S.event man (Events.EnterOnce { once_control;  ran = true }) man in
+        D.join (leave_once first_call) (leave_once later_call)
+      in
+      let is_once = LibraryFunctions.find ~nowarn:true f in
+      match is_once.special args with
+      | Once { once_control; init_routine } -> once once_control init_routine
+      | _
+      | exception LibraryDsl.Expected _ -> (* If there is a mismatch between our library function prototype and the call here, this is raised. However, we catch this here to still do `tf_special_call` where this case is handeled *)
```

Comment:
If our library function specification is wrong, we should fix it. So I'm not sure why we should be handling this here now.
Previously the exception wasn't even exposed, so it wasn't treated specially anywhere else and that wasn't an issue.
Also, if `LibraryFunctions` raises this exception on `f` here and we delegate to the analyses with `S.special`, many of them will also just make the same library function lookup and get the same exception, which isn't caught and crashes Goblint.

The comment doesn't make it any more clear to me. It says

> However, we catch this here to still do `tf_special_call` where this case is handeled

but this whole thing is in `tf_special_call` to begin with.

Reply:
The idea was to delegate the call to `S.special` to get the same behavior and warning as before. An  exception being raised in the analysis where the huge pattern match over all possible library functions is, is arguably more useful to a developer then raising it here, leading them on a wild goose chase about `pthread_once` that has nothing to do with the problem.

Reply:
But I don't have strong opinions, I would also be fine with raising on the top level.

Reply:
I think I would then prefer not exposing the exception at all. We can just have a comment the code here to note that such exceptions raised from here are probably unrelated to `pthread_once`.